Create Family_tree (F_tree)

Alex5202 opened the pull request:
Этот код демонстрирует основы работы с генеалогическим древом:
1.	Класс Person представляет человека с его основными атрибутами и отношениями.
2.	Класс FamilyTree хранит всех людей в дереве.
3.	Класс Research позволяет проводить исследования, в данном случае - находить детей заданного человека.
4.	В Main классе мы создаем пример семейного дерева и демонстрируем, как найти детей конкретного человека.

Commit:
first commit

Commit:
create Home work1 "Family_tree"

Commit:
create Serializable

Alex5202 commented:
Теперь приложение может сохранять семейное древо в файл и загружать его обратно, что позволяет сохранять данные между запусками программы.

Liberate520 commented on src/F_tree/Person.java:
```diff
@@ -0,0 +1,47 @@
+
+public class Person implements Serializable {
+    private String name;
+    private String gender;
```
лучше использовать enum

Liberate520 commented on src/F_tree/Person.java:
```diff
@@ -0,0 +1,47 @@
+        return children;
+    }
+
+    @Override
+    public String toString() {
+        return name;
+    }
```
слишком мало для метода toString. По сути он ничем не отличается от метода getName

Commit:
create Serializable

Alex5202 reviewed (commented):
Внесена рекомендация:
Использует enum для описания гендера.
Сделано ДЗ №3
Класс FamilyTree реализует интерфейс Iterable.
Реализованы методы сортировки по имени и дате рождения.
Создана пакетная структура для проекта.

Commit:
upgraded project with a family tree.
Change the tree, make the class parameterized. It can be used not only for people

Alex5202 commented:
Теперь  генеалогическое древо стало параметризированным и может работать с любым классом, реализующим интерфейс TreeMember. Это позволяет использовать его не только для людей, но и для других сущностей, например, для собак, просто создав новый класс, реализующий интерфейс TreeMembe

Commit:
Added a console application for working with the family tree, implemented using the MVP pattern

Alex5202 commented:
Домашнее задание 4.
Теперь наш проект реализует паттерн MVP:
Model: Класс FamilyTree и связанные с ним классы в пакете model.
View: Класс ConsoleView в пакете view, отвечающий за взаимодействие с пользователем через консоль.
Presenter: Класс FamilyTreePresenter в пакете presenter, который связывает Model и View, обрабатывает пользовательский ввод и обновляет данные.

Liberate520 commented on src/familytree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,119 @@
+public class FamilyTreePresenter {
+    private FamilyTree<Person> model;
+    private ConsoleView view;
+    private FileHandler<Person> fileHandler;
```
нужно создать класс Сервис для объединения работы модели

Liberate520 commented on src/familytree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    private void addPerson() {
+        Person person = view.getPersonInfo();
```
View не должно создавать объекты модели (людей)

Liberate520 commented on src/familytree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    private Person findPersonByName(String name) {
+        for (Person person : model) {
+            if (person.getName().equalsIgnoreCase(name)) {
+                return person;
+            }
+        }
+        return null;
```
за работу с информацией должна отвечать модель

Liberate520 commented on src/familytree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,119 @@
+        showAllPeople();
+    }
+
+    private void saveToFile() {
+        String fileName = view.getFileName();
+        try {
+            fileHandler.save(model, fileName);
+            view.showMessage("Family tree saved successfully.");
+        } catch (Exception e) {
+            view.showMessage("Error saving family tree: " + e.getMessage());
+        }
+    }
+
+    private void loadFromFile() {
+        String fileName = view.getFileName();
+        try {
+            FamilyTree<Person> loadedTree = fileHandler.load(fileName);
+            model = loadedTree;
+            view.showMessage("Family tree loaded successfully.");
+        } catch (Exception e) {
+            view.showMessage("Error loading family tree: " + e.getMessage());
+        }
+    }
```
Реализовать через обращение к Сервис

Commit:
This project structure follows the principles of SOLID

Liberate520 commented on src/familytree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,106 @@
+            view.showMenu();
+            int choice = view.getUserChoice();
+
+            switch (choice) {
+                case 1:
+                    addPerson();
+                    break;
+                case 2:
+                    showAllPeople();
+                    break;
```
свич кейс нарушает второй принцип солид

Liberate520 commented on src/familytree/presenter/FamilyTreePresenter.java:
```diff
@@ -0,0 +1,106 @@
+        String name = view.getPersonName();
+        LocalDate birthDate = view.getPersonBirthDate();
+        String gender = view.getPersonGender();
+        Person person = new Person(name, birthDate, Gender.valueOf(gender));
```
презентер не должен создавать объекты модели (людей). Это нарушает первый принцип солид

Liberate520 commented on src/familytree/service/FamilyTreeServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+
+    public FamilyTreeServiceImpl() {
+        this.familyTree = new FamilyTree<>();
+        this.fileService = new FileService();
```
прямая зависимость нарушает пятый принцип солид

Liberate520 commented on src/familytree/view/ConsoleView.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.List;
+import java.util.Scanner;
+
+public class ConsoleView {
```
должен быть интерфейс в соответствии с пятым принципом солид

Liberate520 commented on src/familytree/view/ConsoleView.java:
```diff
@@ -0,0 +1,71 @@
+        this.scanner = new Scanner(System.in);
+    }
+
+    public void showMenu() {
+        System.out.println("\n--- Family Tree Menu ---");
+        System.out.println("1. Add person");
+        System.out.println("2. Show all people");
```
пункты меню нарушают второй принцип солид
